### PR TITLE
16 add wallet management

### PR DESCRIPTION
**Pull request by gregLibert**

### CODE REVIEW

**Comment by gregLibert** [comment]
@jwmdev this branch is about implementing the wallet management API. In the issue #24 the idea is to work on the GUI wrapping this API. But first I need to finish and merge this issue :-)

File: api/swagger/swagger.yml [hunk 1]
```diff
@@ -1,16 +1,134 @@
...
+      produces:
+        - application/json
+      responses:
+        '200':
+          description: Wallets retrieved
+          schema:
+            type: array
+            items:
+              $ref: '#/definitions/Wallet'
+        '400':
```
Comment (gregLibert) [inline api/swagger/swagger.yml:26]: No 404 HTTP Code ?

File: api/swagger/swagger.yml [hunk 2]
```diff
@@ -1,16 +1,134 @@
...
 schemes:
-- http
-- https
-swagger: "2.0"
+  - http
+  - https
+swagger: '2.0'
 paths:
+  /mgmt/wallet:
+    get:
```
Comment (gregLibert) [inline api/swagger/swagger.yml:15]: I guess a getter for a specific id will be need at some point to extract the wallet.
New _good first issue_ ?

File: api/swagger/swagger.yml [hunk 3]
```diff
@@ -1,16 +1,134 @@
...
+          schema:
+            type: array
+            items:
+              $ref: '#/definitions/Wallet'
+        '400':
+          description: Bad request.
+          schema:
+            $ref: '#/definitions/Error'
+        '422':
+          description: Unprocessable Entity - wallet.json file is corrupted.
```
Comment (gregLibert) [inline api/swagger/swagger.yml:31]: This looks very specific. Which means that when we will use this HTTP code for something else, we will need to update the swagger file...

File: cmd/thyra-server/main.go [hunk 4]
```diff
@@ -28,8 +32,8 @@ func main() {
 		}
 	}()
 
-	server.Port = 80
-	server.TLSPort = 443
+	server.Port = 8080
```
Comment (gregLibert) [inline cmd/thyra-server/main.go:35]: No, I think that at the main branch level, to ensure an easier usage to enduser, we should keep 80 and 443 ports.
But we can talk about putting an environment variable to cover this usage.

Good first issue ?

File: cmd/thyra-server/wallet.json [hunk 5]
```diff
@@ -0,0 +1,44 @@
+[
```
Comment (gregLibert) [inline cmd/thyra-server/wallet.json:1]: To delete

File: cmd/wallet/main.go [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+func main() {
+	w, err := wallet.New("testing")
+	if err != nil {
+		panic(err)
+	}
+
+	fmt.Println("init", w.KeyPairs[0].PrivateKey)
+
+	// this panic because serialization of unprotected keypair is forbidden
```
Comment (gregLibert) [inline cmd/wallet/main.go:17]: dead code to be removed

File: int/apihandler/cmd/execute_function.go [hunk 7]
```diff
@@ -1,46 +1,54 @@
...
 	}
 
 	addr = addr[1:]
 
-	if *params.Body.KeyID != "default" {
+	value, ok := f.walletStorage.Load(*params.Body.KeyID)
+	if !ok {
 		e := errorCodeUnknownKeyID
-		msg := "Error: unknown key id (" + *params.Body.KeyID + ")"
+		msg := "Error: unknown wallet id: " + *params.Body.KeyID
```
Comment (gregLibert) [inline int/apihandler/cmd/execute_function.go:35]: Is it a keyId or a nickname ?

File: int/apihandler/cmd/execute_function.go [hunk 8]
```diff
@@ -1,46 +1,54 @@
...
 				Message: &msg,
 			})
 	}
 
-	pubKey, err := base58.CheckDecode("zkTGqfwJp43tY4FPgRXC7fr2xML3kDQ8bch15SpnDehuxWiKS")
-	if err != nil {
-		panic(err)
+	storedWallet, ok := value.(*wallet.Wallet)
+	if !ok {
+		panic("stored value is not a wallet")
```
Comment (gregLibert) [inline int/apihandler/cmd/execute_function.go:46]: We can leave it like that, but we should create a new issue specifying that the panic should be removed.

File: pkg/front/fileGenerator.go [hunk 9]
```diff
@@ -0,0 +1,260 @@
+//
```
Comment (gregLibert) [inline pkg/front/fileGenerator.go:1]: Why do we need this file ?

File: pkg/wallet/wallet.go [hunk 10]
```diff
@@ -101,6 +113,27 @@ func FromYAML(raw []byte) (w Wallet, err error) {
 	return
 }
 
+func readWallets() (wallets []Wallet) {
+	bytesInput, e := ioutil.ReadFile("wallet.json")
```
Comment (gregLibert) [inline pkg/wallet/wallet.go:117]: Create a new issue to speciy that each wallet should be in its own file.

File: pkg/wallet/wallet.go [hunk 11]
```diff
@@ -133,5 +166,62 @@ func New(nickname string) (*Wallet, error) {
...
+	return &wallet, nil
+}
+
+func Update(wallet Wallet) (err error) {
+	wallets := readWallets()
+
+	wallets = append(wallets, wallet)
+	bytesOutput, err := json.Marshal(wallets)
+	if err != nil {
+		panic(err)
```
Comment (gregLibert) [inline pkg/wallet/wallet.go:186]: why not returning the error instead of panicking ?

File: pkg/wallet/wallet.go [hunk 12]
```diff
@@ -133,5 +166,62 @@ func New(nickname string) (*Wallet, error) {
...
+	}
+
+	os.WriteFile("wallet.json", bytesOutput, 0644)
+	return err
```
Comment (gregLibert) [inline pkg/wallet/wallet.go:190]: no, you should returned nil.

File: pkg/wallet/wallet.go [hunk 13]
```diff
@@ -133,5 +166,62 @@ func New(nickname string) (*Wallet, error) {
 			Salt:       salt,
 			Nonce:      nonce,
 		}},
-	}, nil
+	}
+
+	wallets := readWallets()
+	wallets = append(wallets, wallet)
+
+	bytesOutput, err := json.Marshal(wallets)
```
Comment (gregLibert) [inline pkg/wallet/wallet.go:174]: the error is never checked

File: pkg/wallet/wallet.go [hunk 14]
```diff
@@ -101,6 +113,27 @@ func FromYAML(raw []byte) (w Wallet, err error) {
 	return
 }
 
+func readWallets() (wallets []Wallet) {
+	bytesInput, e := ioutil.ReadFile("wallet.json")
+	wallets = []Wallet{}
+	if e != nil {
+		fmt.Print("No wallet dectected, new one created")
+	} else {
+		json.Unmarshal(bytesInput, &wallets)
```
Comment (gregLibert) [inline pkg/wallet/wallet.go:122]: this function returns an error. You should check it.

File: pkg/wallet/wallet.go [hunk 15]
```diff
@@ -101,6 +113,27 @@ func FromYAML(raw []byte) (w Wallet, err error) {
...
+	}
+	return wallets
+
+}
+
+func GetWallets(w http.ResponseWriter, r *http.Request) ([]Wallet, error) {
+	bytesInput, e := ioutil.ReadFile("wallet.json")
+
+	w.Header().Set("Content-Type", "application/json")
+	w.Write([]byte(bytesInput))
```
Comment (gregLibert) [inline pkg/wallet/wallet.go:132]: this function returns an error. You should check it.

File: pkg/wallet/wallet.go [hunk 16]
```diff
@@ -133,5 +166,62 @@ func New(nickname string) (*Wallet, error) {
...
+func Update(wallet Wallet) (err error) {
+	wallets := readWallets()
+
+	wallets = append(wallets, wallet)
+	bytesOutput, err := json.Marshal(wallets)
+	if err != nil {
+		panic(err)
+	}
+
+	os.WriteFile("wallet.json", bytesOutput, 0644)
```
Comment (gregLibert) [inline pkg/wallet/wallet.go:189]: this function returns an error. You should check it.

File: pkg/wallet/wallet.go [hunk 17]
```diff
@@ -133,5 +166,62 @@ func New(nickname string) (*Wallet, error) {
...
+	wallets := readWallets()
+
+	for index, element := range wallets {
+		if element.Nickname == nickname {
+			wallets = append(wallets[:index], wallets[index+1:]...)
+		}
+	}
+
+	bytesOutput, err := json.Marshal(wallets)
+	os.WriteFile("wallet.json", bytesOutput, 0644)
```
Comment (gregLibert) [inline pkg/wallet/wallet.go:203]: this function returns an error. You should check it.

File: pkg/wallet/wallet.go [hunk 18]
```diff
@@ -133,5 +166,62 @@ func New(nickname string) (*Wallet, error) {
...
+		w.Header().Set("Content-Type", "application/json")
+	} else if strings.Index(target, ".html") > 0 {
+		fileText = front.WalletHtml
+		w.Header().Set("Content-Type", "text/html")
+	} else if strings.Index(target, ".webp") > 0 {
+		fileText = front.Logo_massaWebp
+		w.Header().Set("Content-Type", "image/webp")
+	}
+
+	w.Write([]byte(fileText))
```
Comment (gregLibert) [inline pkg/wallet/wallet.go:226]: this function returns an error. You should check it.

**Comment by Trottoir** [comment]
I'm a Test

File: api/swagger/swagger.yml [hunk 19]
```diff
@@ -1,16 +1,134 @@
...
+          schema:
+            $ref: '#/definitions/Wallet'
+        '400':
+          description: Bad request.
+          schema:
+            $ref: '#/definitions/Error'
+        '422':
+          description: Unprocessable Entity - syntax is correct, but the server was unable to process the contained instructions.
+          schema:
+            $ref: '#/definitions/Error'
```
Comment (Trottoir) [inline api/swagger/swagger.yml:70]: Test

Comment (gregLibert) [reply api/swagger/swagger.yml:70]: yes

File: cmd/thyra-server/main.go (hunk 4 above)
Comment (Trottoir) [reply cmd/thyra-server/main.go:35]: Okay, a flag can be a good idea.
By default 80 else take the flagged value

File: pkg/front/fileGenerator.go (hunk 9 above)
Comment (Trottoir) [reply pkg/front/fileGenerator.go:1]: @gregLibert In order to generate HTML files on running the app

In fact, in dev mode, each time I run the `go run main.go`, it generates the html, css & js files through this : 
https://github.com/massalabs/thyra/blob/95b2468a9872a2c1e6947bd615cf0bb381007b2b/cmd/thyra-server/main.go#L19

File: int/apihandler/cmd/execute_function.go (hunk 7 above)
Comment (Trottoir) [reply int/apihandler/cmd/execute_function.go:35]: Nickname is better

**Pull request merged**